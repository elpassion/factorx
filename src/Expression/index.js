// @flow

import Position from '../Position';

export default class Expression {
  value: string;
  position: Position;

  constructor(value: string, start: number, end: number) {
    this.value = value;
    this.position = new Position(start, end);
  }

  static fromNode(node: Object, code: string): Expression {
    return new Expression(code.slice(node.start, node.end), node.start, node.end);
  }
}
