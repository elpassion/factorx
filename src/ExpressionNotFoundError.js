// @flow

export default class ExpressionNotFoundError extends Error {
  constructor(message?: string) {
    super(message);
    this.name = 'ExpressionNotFoundError';
  }
}
