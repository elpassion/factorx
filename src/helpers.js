import type {selection} from './types'

export function normalizeSelection ({start, end}: selection): selection {
  return { start: changeLocationLine(start, 1), end: changeLocationLine(end, 1) }
}

export function denormalizeSelection ({start, end}: selection): selection {
  return { start: changeLocationLine(start, -1), end: changeLocationLine(end, -1) }
}

function changeLocationLine ({line, column}, diff) {
  return { line: line + diff, column }
}
