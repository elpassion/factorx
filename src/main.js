// @flow
const cst = require('cst')
const {
  VariableDeclarator,
  VariableDeclaration,
  Identifier
} = cst.types
const {Token} = cst

export function tokenAt (ast: Object, line: number, column: number) {
  function isInTokenColumnRange (token) {
    const {start, end} = token.getLoc()
    return start.column <= column && end.column >= column
  }

  function isInTokenLineRange (token) {
    return token.getLoc().start.line === line
  }

  let token = ast.getFirstToken()

  while (!isInTokenLineRange(token) || !isInTokenColumnRange(token)) {
    token = token.getNextToken()
  }
  return token
}

export function extractVariable (
  code: string,
  line: number,
  column: number,
  variableKind: 'let' | 'const' | 'var',
  variableName: string
) {
  const ast = parse(code)
  const token = tokenAt(ast, line, column)
  const expression = token.parentElement
  let VD = new VariableDeclaration([
    new Token('Keyword', variableKind),
    new Token('Whitespace', ' '),

    new VariableDeclarator([
      new Identifier([
        new Token('Identifier', variableName)
      ]),

      new Token('Whitespace', ' '),
      new Token('Punctuator', '='),
      new Token('Whitespace', ' '),
      expression.cloneElement()
    ]),
    new Token('Whitespace', '\n')
  ])
  expression.parentElement.replaceChild(new Identifier([ new Token('Identifier', variableName) ]), expression)
  ast.prependChild(VD)
  return ast
}

export function parse (code: string): Object {
  return new cst.Parser().parse(code)
}
