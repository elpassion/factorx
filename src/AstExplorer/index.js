// @flow

import * as babel from 'babel-core';
import Position from '../Position';
import Expression from '../Expression';
import ExpressionNotFoundError from '../ExpressionNotFoundError';
import options from './options';

export default class AstExplorer {
  code: string;
  ast: Object;
  map: Object;
  stop: Function;

  constructor(input: string) {
    const { code, map, ast } = babel.transform(input, options);
    this.code = code;
    this.map = map;
    this.ast = ast;
  }

  serializeNode = (node: Object) => Expression.fromNode(this.code, node);

  findExpressions(selection: Position): Array<Expression> {
    const paths = [];
    this.transform(() => ({
      visitor: {
        Expression(path) {
          if (selection.includes(path.node)) {
            paths.push(path);
          }
        },
      },
    }));
    if (paths.length === 0) throw new ExpressionNotFoundError();
    let parentBlock = paths[paths.length - 1].scope.block;
    if (parentBlock.type === 'ArrowFunctionExpression') parentBlock = parentBlock.body;
    return paths
      .filter(path => parentBlock.start <= path.node.start && parentBlock.end >= path.node.end)
      .map(path => path.node)
      .map(this.serializeNode);
  }

  extractVariable(selection: Position): string {
    let extracted = false;
    this.transform(({ types }) => ({
      visitor: {
        Expression: (path) => {
          const { node } = path;
          if (!node.visited && selection.includes(node)) {
            node.visited = true;
            const id = path.scope.generateUidIdentifierBasedOnNode(node.id);
            path.scope.push({ id, init: path.node });
            path.replaceWith(types.identifier(id.name));
            extracted = true;
          }
        },
      },
    }));
    if (!extracted) throw new ExpressionNotFoundError();
    return this.code;
  }

  transform(plugin: Function) {
    const { code, map, ast } = babel.transformFromAst(this.ast, this.code, {
      ...options,
      plugins: [plugin],
    });
    this.code = code;
    this.map = map;
    this.ast = ast;
  }
}
