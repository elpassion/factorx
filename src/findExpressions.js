// @flow
import {Parser} from 'cst'
import type {selection, expression} from './types'
import {flatten} from 'lodash'
import {normalizeSelection, denormalizeSelection} from './helpers'

function parse (code: string): Object {
  return new Parser().parse(code)
}

function findAllExpressions (ast: Object) {
  const nodesIndex = ast._traverse._nodeIndex._index
  return flatten(Object.keys(nodesIndex).map(nodeType =>
    nodesIndex[nodeType].filter(node => node.isExpression)
  ))
}

export function findSelectedExpression (ast: Object, selection: selection) {
  const allExpressions = findAllExpressions(ast)

  function isSearchedExpression (expression) {
    const expressionPosition = expression.getLoc()
    return (
      expressionPosition.start.column === selection.start.column &&
      expressionPosition.start.line === selection.start.line &&
      expressionPosition.end.column === selection.end.column &&
      expressionPosition.end.line === selection.end.line
    )
  }

  return allExpressions.find(isSearchedExpression)
}

function findExpressionsAt (ast: Object, selection: selection) {
  const allExpressions = findAllExpressions(ast)

  function isOneLineExpression (expression) {
    const expressionPosition = expression.getLoc()
    return expressionPosition.start.line === expressionPosition.end.line
  }

  function isOneLineExpressionInRange (expression) {
    const expressionPosition = expression.getLoc()
    return (
      expressionPosition.start.column <= selection.start.column &&
      expressionPosition.start.line === selection.start.line &&
      expressionPosition.end.column >= selection.end.column &&
      expressionPosition.end.line === selection.end.line
    )
  }

  function isMultiLineExpressionInRange (expression) {
    const expressionPosition = expression.getLoc()
    return (
      expressionPosition.start.line <= selection.start.line &&
      expressionPosition.end.line >= selection.end.line
    )
  }

  return allExpressions.filter((expression) => {
    if (isOneLineExpression(expression)) {
      return isOneLineExpressionInRange(expression)
    } else {
      return isMultiLineExpressionInRange(expression)
    }
  }).sort((previousExpression, nextExpression) => {
    const previousRange = previousExpression.getRange()
    const nextRange = nextExpression.getRange()
    return (previousRange[1] - previousRange[0]) - (nextRange[1] - nextRange[0])
  })
}

export function findExpressions (code: string, selection: selection, {depth = 0} : {depth?: number}): Array<expression> {
  const ast = parse(code)
  const normalizedSelection = normalizeSelection(selection)
  const expressions = findExpressionsAt(ast, normalizedSelection).map(expression => ({
    value: expression.getSourceCode(),
    selection: denormalizeSelection(expression.getLoc())
  }))
  if (depth === 0) depth = expressions.length + 1
  return expressions.slice(0, depth)
}
