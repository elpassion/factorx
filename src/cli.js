#! /usr/bin/env node
// @flow
import 'babel-polyfill';
import program from 'commander';
import getStdin from 'get-stdin';
import chunk from 'lodash/chunk';
import { AstExplorer, Position } from '../lib/main';

(() => {
  function writeJSON(message) {
    process.stdout.write(JSON.stringify(message));
  }

  function createMessageFromError({ name, message }) {
    const status = 'error';
    switch (name) {
      case 'ExpressionNotFound': {
        return { status, error: { name, message } };
      }
      case 'SyntaxError': {
        return { status, error: { name, message: 'Unable to parse the code' } };
      }
      default: {
        return { status, error: { name, message } };
      }
    }
  }

  async function extractVariableCmd(selections) {
    const file = await getStdin();
    try {
      const astExplorer = new AstExplorer(file);
      let result;
      if (selections.length === 1) {
        result = astExplorer.extractVariable(selections[0]);
      } else {
        result = astExplorer.extractMultipleVariables(selections);
      }
      writeJSON({ status: 'ok', ...result });
    } catch (error) {
      writeJSON(createMessageFromError(error));
    }
  }

  async function getExpressionsCmd(selection) {
    const file = await getStdin();
    try {
      const astExplorer = new AstExplorer(file);
      const expressions = astExplorer.findExpressions(selection);
      writeJSON({ status: 'ok', expressions });
    } catch (error) {
      writeJSON(createMessageFromError(error));
    }
  }

  async function getExpressionOccurrencesCmd(selection) {
    const file = await getStdin();
    try {
      const astExplorer = new AstExplorer(file);
      const expressions = astExplorer.findExpressionOccurrences(selection);
      writeJSON({ status: 'ok', expressions });
    } catch (error) {
      writeJSON(createMessageFromError(error));
    }
  }

  program
    .command('get-expressions <startPosition> <endPosition>')
    .description('get expressions at range')
    .action((startPosition, endPosition) => {
      const selection = new Position(parseInt(startPosition, 10), parseInt(endPosition, 10));
      getExpressionsCmd(selection);
    });

  program
    .command('extract-variable [positions...]')
    .description('extract variable at range')
    .action((positions) => {
      const intPositions = positions.map(position => parseInt(position, 10));
      const positionPairs = chunk(intPositions, 2);
      const selections = positionPairs.map(([start, end]) => new Position(start, end));
      extractVariableCmd(selections);
    });

  program
    .command('get-expression-occurrences <startPosition> <endPosition>')
    .description('get all expressions of the same value at the same scope')
    .action((startPosition, endPosition) => {
      const selection = new Position(parseInt(startPosition, 10), parseInt(endPosition, 10));
      getExpressionOccurrencesCmd(selection);
    });

  program.version('0.0.1').parse(process.argv);
})();
