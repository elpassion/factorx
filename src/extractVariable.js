// @flow
import type {selection} from './types'
import {Token, Parser, types} from 'cst'
const {VariableDeclarator, VariableDeclaration, Identifier} = types
import {normalizeSelection} from './helpers'
import {findSelectedExpression, ExpressionNotFoundError} from './findExpressions'

function parse (code: string): Object {
  return new Parser().parse(code)
}

function createIdentifier () {
  return new Identifier([new Token('Identifier', '_ref')])
}

function createVariableDeclaration (identifier, expression) {
  return new VariableDeclaration([
    new Token('Keyword', 'const'),
    new Token('Whitespace', ' '),

    new VariableDeclarator([
      createIdentifier(),
      new Token('Whitespace', ' '),
      new Token('Punctuator', '='),
      new Token('Whitespace', ' '),
      expression.cloneElement()
    ]),
    new Token('Whitespace', '\n')
  ])
}

function findExpressionScope (expression) {
  let parent = expression.parentElement
  let child = expression
  while (parent.type !== 'Program' && parent.type !== 'BlockStatement') {
    child = parent
    parent = parent.parentElement
  }
  return {parent, child}
}

export function extractVariable (code: string, selection: selection): string {
  const normalizedSelection = normalizeSelection(selection)
  const ast = parse(code)
  const expression = findSelectedExpression(ast, normalizedSelection)
  if (!expression) throw new ExpressionNotFoundError()
  const identifier = createIdentifier()
  const VD = createVariableDeclaration(identifier, expression)
  const {parent, child} = findExpressionScope(expression)

  parent.insertChildBefore(VD, child)
  expression.parentElement.replaceChild(identifier, expression)

  return ast.getSourceCode()
}
