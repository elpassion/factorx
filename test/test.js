// @flow
import {tokenAt, extractVariable, parse} from '../src/main'

test('tokenAt', () => {
  const code = `5 + 2`
  const ast = parse(code)
  expect(tokenAt(ast, 1, 1).value).toEqual(5)
  expect(tokenAt(ast, 1, 5).value).toEqual(2)

  const code2 = `
5 + 2
const name = 5
  `
  const ast2 = parse(code2)
  expect(tokenAt(ast2, 2, 1).value).toEqual(5)
  expect(tokenAt(ast2, 3, 1).value).toEqual('const')
  expect(tokenAt(ast2, 3, 2).value).toEqual('const')
  expect(tokenAt(ast2, 3, 7).value).toEqual('name')
  expect(tokenAt(ast2, 3, 14).value).toEqual(5)
})

test('extractVariable', () => {
  const code = `5 + 2`
  expect(extractVariable(code, 1, 1, 'let', 'number').getSourceCode()).toMatchSnapshot()
  expect(extractVariable(code, 1, 5, 'let', 'number').getSourceCode()).toMatchSnapshot()
  const code2 = `'abc' + 'cde'`
  expect(extractVariable(code2, 1, 1, 'let', 'string').getSourceCode()).toMatchSnapshot()
  expect(extractVariable(code2, 1, 10, 'const', 'text').getSourceCode()).toMatchSnapshot()
  expect(extractVariable(code2, 1, 10, 'var', 'varText').getSourceCode()).toMatchSnapshot()
  expect(extractVariable(code2, 1, 6, 'let', 'addition').getSourceCode()).toMatchSnapshot()
})
