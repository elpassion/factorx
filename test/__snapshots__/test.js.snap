exports[`test extractVariable 1`] = `
"let number = 5
number + 2"
`;

exports[`test extractVariable 2`] = `
"let number = 2
5 + number"
`;

exports[`test extractVariable 3`] = `
"let string = \'abc\'
string + \'cde\'"
`;

exports[`test extractVariable 4`] = `
"const text = \'cde\'
\'abc\' + text"
`;

exports[`test extractVariable 5`] = `
"var varText = \'cde\'
\'abc\' + varText"
`;

exports[`test extractVariable 6`] = `
"let addition = \'abc\' + \'cde\'
addition"
`;
